File: ifgi/scene/test_Primitive.py
#!/usr/bin/env python
#
# Copyright 2011 (C) Yamauchi, Hitoshi
#
#
# test for Primitive module
#

"""test IFGI Primitive module"""

from PIL import Image           # image

import unittest
import numpy
import random
import Primitive
import Ray
import cProfile
import pstats

# test primitive: triangle ray intersection
def primitive_tri_ray_intersection_sub():
    """primitive: ray-triangle intesection subroutine"""

    iw = 50
    ih = 50
    imgsize = (iw, ih)
    # create a triangle
    xoff = iw/10
    yoff = iw/10
    p0 = numpy.array([xoff,           yoff, -10.0])
    p1 = numpy.array([iw - xoff,      yoff, -10.0])
    p2 = numpy.array([iw/2,      ih - yoff, -10.0])

    tri = Primitive.Triangle()
    tri.set_vertex(p0, p1, p2)

    white   = (255, 255, 255)
    red     = (255,   0,   0)
    resimg = Image.new("RGBA", imgsize, white)

    for x in xrange(0, imgsize[0], 1):
        for y in xrange(0, imgsize[1], 1):
            origin = numpy.array([ x, y, 0])
            dir    = numpy.array([ 0, 0, -1])
            min_t  = 0.1
            max_t  = 100
            r = Ray.Ray(origin, dir, min_t, max_t)
            is_hit, dist = tri.ray_intersect(r)
            if is_hit == True:
                # print origin, 'Hit'

                resimg.putpixel((x, imgsize[1]-y), red)

    resimg.save("res_ray_tri_intersect.png")
    print 'Saved ... res_ray_tri_intersect.png'


# test primitives
class TestPrimitive(unittest.TestCase):
    """test suit for Primitive."""
    # test bbox
    def test_bbox(self):
        """test for bbox"""
        random.seed(0)

        minp = numpy.array([-2.0, -1.0,  1.2])
        maxp = numpy.array([-0.2,  1.0,  5.5])
        ipos = numpy.array([ 0.0,  0.0,  0.0])
        bbox = Primitive.BBox()
        for p in xrange(1000):
            for i in xrange(0, 3, 1):
                ipos[i] = random.uniform(minp[i], maxp[i])
                # print i, ipos
            bbox.insert_point(ipos)

        for i in xrange(2):
            assert(minp[i]     <= bbox.get_min()[i])
            assert(bbox.get_min()[i] <= bbox.get_max()[i])
            assert(bbox.get_max()[i] <= maxp[i])

    # test primitive: one triangle
    def test_primitive_one_tri(self):
        """test primitive: one triangle"""
        tmesh = Primitive.TriMesh()

        vertex_list       = []
        face_idx_list     = []
        texcoord_list     = []
        texcoord_idx_list = []
        normal_list       = []
        normal_idx_list   = []

        # add a triangle
        vertex_list.append(numpy.array([ 1.0, 0.0, 0.0]))
        vertex_list.append(numpy.array([ 0.0, 1.0, 0.0]))
        vertex_list.append(numpy.array([-1.0, 0.0, 0.0]))
        face_idx_list.append(numpy.array([0, 1, 2]))

        tmesh.set_data(vertex_list,
                       face_idx_list,
                       texcoord_list,
                       texcoord_idx_list,
                       normal_list,
                       normal_idx_list
                       )
        bbox = Primitive.BBox()
        bbox.insert_point(numpy.array([-1.0, 0.0, 0.0]))
        bbox.insert_point(numpy.array([ 1.0, 1.0, 0.0]))
        assert(tmesh.get_bbox().equal(bbox))


    # primitive: ray-triangle intesection
    def test_primitive_tri_ray_intersection_sub(self):
        """primitive: ray-triangle intesection"""

        primitive_tri_ray_intersection_sub()

        #--- profile
        # cProfile.run('primitive_tri_ray_intersection_sub()')
        #
        # cProfile.run('primitive_tri_ray_intersection_sub()',
        #              'test_primitive.prof')
        # p = pstats.Stats('test_primitive.prof')

#
# main test
#
suit0   = unittest.TestLoader().loadTestsFromTestCase(TestPrimitive)
alltest = unittest.TestSuite([suit0])
unittest.TextTestRunner(verbosity=2).run(alltest)

File: ifgi/cpp/scene/test_Primitive.cpp
//----------------------------------------------------------------------
// ifgi c++ implementation: test_Primitive.cpp
// Copyright (C) 2010-2012 Yamauchi, Hitoshi
//----------------------------------------------------------------------
/// \file
/// \brief test primitive

#include "Camera.hh"
#include "HitRecord.hh"
#include "ImageFilm.hh"
#include "Ray.hh"
#include "Triangle.hh"

#include <gtest/gtest.h>

using namespace ifgi;

/// Test Primitive, triangle ray intersection test
/// TEST(test_case_name, test_name)
TEST(PrimitiveTest, RayTriIntersectionTest1)
{
    // One ray and one trianle
    // create a triangle
    Scalar_3 p0( 1.0f, 0.0f,    0.0f);
    Scalar_3 p1( 0.0f, 1.7321f, 0.0f);
    Scalar_3 p2(-1.0f, 0.0f,    0.0f);

    Dictionary cam_dict;
    cam_dict.set("cam_name", "default");
    cam_dict.set("eye_pos", "0.0 0.64 0.76");
    cam_dict.set("view_dir", "0.0 0.0 -1.0");
    cam_dict.set("up_dir", "0.0 1.0 0.0");
    cam_dict.set("z_near", "0.1");
    cam_dict.set("z_far", "10000");
    cam_dict.set("resolution_x", "64");
    cam_dict.set("resolution_y", "64");

    Camera cam;
    cam.set_config_dict(cam_dict);

    Triangle tri;
    tri.set_vertex(p0, p1, p2);
    // std::cout << "tri.to_string(): " << tri.to_string() << std::endl;

    HitRecord hr;

    Ray r;
    cam.get_ray(0.5, 0.5, r);
    // std::cout << "Ray: " << r.to_string() << std::endl;

    hr.initialize();
    ASSERT_EQ(tri.ray_intersect(r, hr), true);
}

/// Test Primitive, triangle ray intersection test
/// TEST(test_case_name, test_name)
TEST(PrimitiveTest, RayTriIntersectionTest2)
{
    Scalar const iw = 50.0;
    Scalar const ih = 50.0;
    // imgsize = (iw, ih);

    // create a triangle
    Scalar xoff = iw / 10.0;
    Scalar yoff = iw / 10.0;
    Scalar_3 p0(xoff,           yoff, -10.0);
    Scalar_3 p1(iw - xoff,      yoff, -10.0);
    Scalar_3 p2(iw/2.0,    ih - yoff, -10.0);

    Triangle tri;
    tri.set_vertex(p0, p1, p2);

    Color const white(1.0, 1.0, 1.0, 1.0);
    Color const red  (1.0, 0.0, 0.0, 1.0);
    Sint32_3 const imgsize(static_cast< Sint32 >(iw),
                            static_cast< Sint32 >(ih),
                            4);
    ImageFilm img(imgsize, "RGBA");
    img.fill_color(white);
    Ray r;
    HitRecord hr;

    for(Sint32 x = 0; x < imgsize[0]; ++x){
        for(Sint32 y = 0; y < imgsize[1]; ++y){
            Scalar_3 const origin(static_cast< Scalar >(x),
                                  static_cast< Scalar >(y),
                                  0.0);
            Scalar_3 const dir(0.0, 0.0, -1.0);
            Scalar const min_t = 0.1;
            Scalar const max_t = 100.0;
            r.initialize(origin, dir, min_t, max_t);
            hr.initialize();
            if(tri.ray_intersect(r, hr)){
                // std::cout << "test_Primitive: Hit: " << origin << std::endl;
                img.put_color(x, y, red);
            }
        }
    }

    std::string const fname = "res_ray_tri_intersect.ppm";
    img.save_file(fname, "ppm");
    std::cout << "Saved ... [" << fname << "]" << std::endl;
}

/// Test Primitive, triangle ray intersection test
/// TEST(test_case_name, test_name)
TEST(PrimitiveTest, BBoxTest)
{
    srand48(0);

    Scalar_3 const minp(-2.0, -1.0,  1.2);
    Scalar_3 const maxp(-0.2,  1.0,  5.5);
    Scalar_3 ipos( 0.0,  0.0,  0.0);
    BBoxScalar bbox;
    for(Sint32 p = 0; p < 1000; ++p){
        for(Sint32 i = 0; i < 3; ++i){
            ipos[i] = drand48() * (maxp[i] - minp[i]) + minp[i];
            // std::cout << i << ", " << ipos << std::endl;
        }
        bbox.insert_point(ipos);
    }
    for(Sint32 i = 0; i < 3; ++i){
        // std::cout << "[" << bbox.get_min() << "]["
        //           << bbox.get_max() << "]" << std::endl;
        ASSERT_LE(minp[i],           bbox.get_min()[i]);
        ASSERT_LE(bbox.get_min()[i], bbox.get_max()[i]);
        ASSERT_LE(bbox.get_max()[i], maxp[i]);
    }
}

int main(int argc, char **argv)
{
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}



